### PR TITLE
add bulk api and pp method

### PR DESCRIPTION
**Pull request by Datseris**
I am putting in the same file my suggestion for the bulk api as well as the PP method.

I've tested both in my scientific project and they worked like a charm.

If you like this, then I'd say to move the main API to a central file, and extend it to the other methods in a _different_ PR.

### CODE REVIEW

**Comment by kahaaga** [comment]
> I've tested both in my scientific project and they worked like a charm.

Nice! Have you tried to reproduce any plots from the paper? Would be nice to see that we get roughly the same thing on the Rössler system that they use.

> I am putting in the same file my suggestion for the bulk api as well as the PP method.

> If you like this, then I'd say to move the main API to a central file, and extend it to the other methods in a _different_ PR.

Agreed. Move the main API to a  `SurrogateGenerator.jl` file, merge this PR, and then open a new PR that extends it to the other methods?

**Comment by kahaaga** [comment]
@Datseris Before merging this, a test that `surrogenerator(x::AbstractVector, pp::PseudoPeriodic)` gives a valid time series should be added. 

Does the PPS surrogate contain a shuffled version of the original signal (constrained surrogate)? If so, a test `@test sort(x) .== sort(s)` should be added. If the surrogate does not contain precisely the same values, is there another clever way of making sure the method works?

**Comment by Datseris** [comment]
I was **not** able to reproduce results from the paper. I don't know how the calculate the correlation dimension, but I use DynamicalSystems.jl and compute the generalized dimension with q=2, which is supposed to be equivalent.

```julia
using TimeseriesSurrogates, DynamicalSystems

ds = Systems.roessler(; b=2, c =4, a=0.309)
tr = trajectory(ds, 5000*0.2; dt=0.2, Ttr = 10)
x = tr[:, 1] .+ randn(5001)*0.05
τ = 9 # delay time = when AC becomes zero
ρ = 0.02
d = 3 # embedding dimension
shift = true

pp = PseudoPeriodic(d, τ, ρ, shift)
sg = surrogenerator(x, pp)

using PyPlot
figure()
εs = ℯ .^ (-2.5:0.05:-0.5)
H = genentropy.(2, εs, Ref(sg.init.z))
plot(log.(εs), H)
for _ in 1:30
    s = sg()
    H = genentropy.(2, εs, Ref(sg.init.y))
    plot(log.(εs), H, color = "k", ls = "dotted", alpha = 0.2)
end
xlabel("log(ε)")
ylabel("H₂")
```

gives

image


which should be like their figure 4, right panel. 

Notice that his work is **extremely fishy** from an application perspective. The algorithm is great, but how they apply it seems very poor scientifically for me. If you look at their figure 5 (b), they say that the result of the original data is "different" from surrogates. But the difference is only at very small ε, the correlation dimension (the slope of the curve) is too similar to claim difference.

I wonder how they got such a subjective and vague statement into PRL. I was even thinking of sending them an email with my reproduction, that shows that the random roessler considered here falls in their "correlated noise" category.

In general I don't bother that much reproducing such papers, because the authors always make tremendously poor job describing **exactly how** they make their figures.

**Comment by Datseris** [comment]
ah, I also used wrong ρ, giving `ρ=0.0025` as they do in the paper gives:

image


Notice that the statement about the vaguety of how to reject the null is still very valid. How much similar should the correlation dimension be to reject it? this is super vague.

**Comment by kahaaga** [comment]
> If you look at their figure 5 (b), they say that the result of the original data is "different" from surrogates. But the difference is only at very small ε, the correlation dimension (the slope of the curve) is too similar to claim difference.

As a hard-rock geologist trained to use my eyes and instincts to deduce properties of the natural world, I find a casual "it's different" welcoming... buut it may not be the most robust of determining the validity of the method :laughing:

One possible way to determine whether the surrogates and the original have different generalized dimension is to:
- Calculate the generalized dimension for smaller chunks of the original time series, giving an ensemble of dimension estimates.
- Calculate the generalized dimension for an ensemble of surrogate time series.
- Ask: does the median of the surrogate dimension distribution lie outside some quantile of the dimension estimates for the original signal? (both upper and lower quantiles must be checked). If so, one claim claim a "difference" that is quantifiable. 
- One could also make a formal hypothesis that checks that the distributions for the dimension of the original signal vs. the distribution of dimensions for are different.

> I was even thinking of sending them an email with my reproduction, that shows that the random roessler considered here falls in their "correlated noise" category.

Yes, it is always annoying when analysis parameters are not stated properly. It might very well be that for some parameters, the Rössler system falls in the correlated noise category, while for some parameters it does not. The only way to know is to  do an ensemble analysis where for each set of parameters for the Rössler system, you do some ensemble comparison like described above.

**Comment by Datseris** [comment]
I've addded some basic tests for PseudoPeriodic, but I would argue that we shouldn't spent too much time at the tests now, since many will be rewritten and it feels counter productive. 

I've sketched a test of how to test the "actual" algorithm, outside length, but I would prefer to do that once we have written everything in the new version.

**Comment by Datseris** [comment]
we also need to do a bit of tidying of the source :D I count way too many files in the `methods` folder given how many methods we actually have :D

**Comment by Datseris** [comment]
To give an example of the new interface: e.g. for `RandomFourier`, the surrogate method now becomes just 
```
RandomFourier(phases=true)
```
because the two fields `forward, inverse` will become part of the `init` field of the `SurrogateGenerator`.

**Comment by Datseris** [comment]
Although, we should make a decision on the bulk api. There are two ways to do it:

1. What I've written here, where we have the `Generator` thing.
2. The second way is that every `<: Surrogate` struct also contains the initialized things that they want, and there is only a single function `surrogate(x, method::Surrogate)`, since now the moment you type e.g. `PseudoPeriodic(a,b,c)`, this instantiation also takes care of initializing everything.

I think number 2 is simpler...? So for `PseudoPeriodic`, the things I've written as `init = (y = y, w = w, z = z)` , well, actually the `y, w, z` will become a field of `PseudoPeriodic`. This way we eliminate the `SurrogateGenerator` struct completely.

What do you think?

**Comment by Datseris** [comment]
Aaaaaah sorry I am sorry, I **now remember why I did this Generator thing**. Sorry, we can't do (2) even though it does sound simpler...

For some methods, like `PseudoPeriodic`, the input timeseries `x` is crucial for the preinitialization. Thus, we can't efficiently loop over `surrogate(x, method)` because `x` must be known to pre-initialize method.

**Comment by Datseris** [comment]
Notice that this also means that in e,g. `RandomFourier` we **do not need to calculate teh fourier transform of x everytime**: we only need to do it once and then simply modify it!

File: src/methods/pseudoperiodic.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+export PseudoPeriodic, noiseradius
+"""
+    PseudoPeriodic(d, τ, ρ, shift=true) <: Surrogate
+Create surrogates suitable for pseudo-periodic signals. They retain the periodic structure
+of the signal, while inter-cycle dynamics that are either deterministic or correlated
+noise are destroyed (for appropriate `ρ` choice).
+Therefore these surrogates are suitable to test the null hypothesis
+that the signal is periodic with uncorrelated noise[^Small2001].
+
+Argumensts `d, τ, ρ` are as in the paper, the embedding dimension, delay time and
```
Comment (kahaaga) [inline src/methods/pseudoperiodic.jl:11]: Small typo: "Arguments" instead of "Argumensts"

**Pull request merged**